DownFall/TargetingViewModel.swift

```diff
@@ -0,0 +1,190 @@
+//
+//  TargetingViewModel.swift
+//  DownFall
+//
+//  Created by Katz, Billy on 1/22/20.
+//  Copyright © 2020 William Katz LLC. All rights reserved.
+//
+
+import Foundation
+
+struct Target {
+    let coord: TileCoord
+    let isLegal: Bool
+}
+
+protocol TargetingOutputs {
+    var toastMessage: String { get }
+    var currentTargets: [Target] { get }
+    var legallyTargeted: Bool { get }
+    
+}
+
+protocol TargetingInputs {
+    
+    /// Use this to choose targets
+    func didTarget(_ coord: TileCoord)
+    
+    /// Use this to consume the item
+    func didUse()
+}
+
+protocol Targeting: TargetingOutputs, TargetingInputs {}
+
+class TargetingViewModel: Targeting {
+    
+    public var updateCallback: (() -> Void)?
+    
+    var ability: AnyAbility? {
+        didSet {
+            
+            currentTargets = []
+            
+            if ability == nil {
+                InputQueue.append(Input(InputType.itemUseCanceled))
+            } else {
+                if oldValue == nil {
+                    InputQueue.append(Input(InputType.itemUseSelected(ability!)))
+                }
+                autoTarget()
+            }
+            updateCallback?()
+        }
+    }
+    
+    var numberOfTargets: Int {
+        return ability?.targets ?? 0
+    }
+    
+    var typesOfTargets: [TileType] {
+        return ability?.targetTypes ?? []
+    }
+    
+    var toastMessage: String {
+        if ability == nil {
+            return ""
+        }
+        let baseString = "Choose "
+        let types = Set<String>(self.typesOfTargets.map {
+            return $0.humanReadable
+        }).joined(separator: ",")
+        return baseString + "\(self.numberOfTargets) " + types
+    }
+    
+    var currentTargets: [Target] = [] {
+        didSet {
+            self.updateCallback?()
+            InputQueue.append(Input(InputType.itemCanBeUsed(legallyTargeted)))
+        }
+    }
+    
+    
+    var legallyTargeted: Bool {
+        return self.currentTargets.allSatisfy({
+            return $0.isLegal
+        }) && self.currentTargets.count == self.numberOfTargets
+    }
+    
+    
+    private var tiles: [[Tile]]? = nil {
+        didSet {
+            autoTarget()
+        }
+    }
+    
+    init() {
+        Dispatch.shared.register { [weak self] (input) in
+            switch input.type {
+            case .itemUseSelected(let ability):
+                self?.ability = ability
+                self?.updateCallback?()
+            case .transformation(let trans):
+                if let inputType = trans.inputType,
+                    case InputType.itemUseSelected(_) = inputType,
+                    let endTiles = trans.endTiles
+                {
+                    self?.tiles = endTiles
+                    self?.updateCallback?()
```

Comment:
This call is spewn around the code.  Can we consolidate where this should be called? I think in didSets is good